Sources/GRPC/GRPCServerPipelineConfigurator.swift

```diff
@@ -0,0 +1,350 @@
+          context.close(mode: .all, promise: nil)
+        } else {
+          self.configuration.logger.warning("No ALPN protocol negotiated'")
+        }
+      }
+
+    case .shutdownCompleted:
+      // We don't care about this here.
+      ()
+    }
+  }
+
+  // MARK: - Channel Handler
+
+  internal func errorCaught(context: ChannelHandlerContext, error: Error) {
+    if let delegate = self.configuration.errorDelegate {
+      let baseError: Error
+
+      if let errorWithContext = error as? GRPCError.WithContext {
+        baseError = errorWithContext.error
+      } else {
+        baseError = error
+      }
+
+      delegate.observeLibraryError(baseError)
+    }
+
+    context.close(mode: .all, promise: nil)
+  }
+
+  internal func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    switch self.state {
+    case .notConfigured:
+      if let event = event as? TLSUserEvent {
+        self.handleHandshakeCompletedEvent(event, context: context)
+      }
+    case .configuring:
+      ()
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  internal func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    self.bufferedReads.append(data)
+
+    switch self.state {
+    case .notConfigured:
+      let buffer = self.unwrapInboundIn(data)
+      self.determineHTTPVersionAndConfigurePipeline(usingBuffer: buffer, context: context)
+
+    case .configuring:
+      // Just buffer, that's okay.
+      ()
+    }
+
+    // Don't forward the reads: we'll do so when we have configured the pipeline.
+  }
+
+  internal func removeHandler(
+    context: ChannelHandlerContext,
+    removalToken: ChannelHandlerContext.RemovalToken
+  ) {
+    // Forward any buffered reads.
+    while let read = self.bufferedReads.popFirst() {
+      context.fireChannelRead(read)
+    }
+    context.leavePipeline(removalToken: removalToken)
+  }
+}
+
+// MARK: - HTTP Version Parser
+
+struct HTTPVersionParser {
+  /// HTTP/2 connection preface bytes. See RFC 7540 § 5.3.
+  private static let http2ClientMagic = [
+    UInt8(ascii: "P"),
+    UInt8(ascii: "R"),
+    UInt8(ascii: "I"),
+    UInt8(ascii: " "),
+    UInt8(ascii: "*"),
+    UInt8(ascii: " "),
+    UInt8(ascii: "H"),
+    UInt8(ascii: "T"),
+    UInt8(ascii: "T"),
+    UInt8(ascii: "P"),
+    UInt8(ascii: "/"),
+    UInt8(ascii: "2"),
+    UInt8(ascii: "."),
+    UInt8(ascii: "0"),
+    UInt8(ascii: "\r"),
+    UInt8(ascii: "\n"),
+    UInt8(ascii: "\r"),
+    UInt8(ascii: "\n"),
+    UInt8(ascii: "S"),
+    UInt8(ascii: "M"),
+    UInt8(ascii: "\r"),
+    UInt8(ascii: "\n"),
+    UInt8(ascii: "\r"),
+    UInt8(ascii: "\n"),
+  ]
+
+  /// Determines whether the bytes in the `ByteBuffer` are prefixed with the HTTP/2 client
+  /// connection preface.
+  static func prefixedWithHTTP2ConnectionPreface(_ buffer: ByteBuffer) -> Bool {
+    var buffer = buffer
+
+    guard let bytes = buffer.readBytes(length: HTTPVersionParser.http2ClientMagic.count) else {
```

Comment:
It's a shame to allocate an array here. You can use `ByteBufferView.elementsEqual` to get something pretty close, I think, that skips the alloc.

Reply:
Oh neat, I wasn't aware of that one.